- Open and read the txt file

In [1]:
with open("data/chapter2/the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
print("Total number of character:", len(raw_text))
print(raw_text[:99])


Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


- Here we open and read the file.
- The file has 20479 characters

# Converting text to tokens.

In [2]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text) # Using regex to split the word on spaces
print(result)


['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']


- Note that even  space is considered here. Another thing to note is that commas and periods are considered part of the word, and hence have to be separated
- Modify to remove commas and periods.

In [3]:
result = re.split(r'([,.]|\s)', text)
print(result)

['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ',', '', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


- Since there are white spaces after every word, it better to remove them. This will save only the words and the punctuations.
- Remove whitespace

In [4]:
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


- Above is an example of how to handle only commas and periods.
- But we need a way to handle all types of punctuations.

In [5]:
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)', text) # Regex for all types of punctuation.
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


- apply to the entire raw_text. This way the entire raw text is now processed.
- This will separate the words and punctuations as separate tokens and will also remove the which spaces.


In [6]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [item for item in preprocessed if item.strip()]
print(len(preprocessed))

4690


- After the processing there are 4690 tokens. Example of the first 30 tokens is printed below.

In [7]:
print(preprocessed[:30])

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in']


- Since the computer dont function with text, we will give each word/token an ID.
- This way each word will be represented by its ID.
- Below is the logic to assign an ID to each token.
    - First the words are sorted (alphanumerically)
    - After sorting each word is given a number based on its order.

In [8]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


- We can see that there are 1130 distinct words/tokens in our vocabulary.
- So we will create a mapping dictionary to convert them to numbers.
- The dictionary will have 1130 items starting from 0 to 1129.


In [9]:
vocab = {token:integer for integer, token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i>50:
        break

('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)
('His', 51)


# Implement a simple text tokenizer
- It should convert tokens to integers and back.
- Basically, it stores the token -> int and int -> token mappings.
- It is used to encode and decode text.

In [10]:
class SimpleTokenizerV1:
    '''
    A simple class to tokenize text.
    Input is the dictionary of {str: int}
    Creates the reverese mapping -> {int: str}
    Used for encoding string to integer input and decoding the integer back to string.
    '''
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [item for item in preprocessed if item.strip()]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        text = ' '.join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

# Example

In [11]:
tokenizer = SimpleTokenizerV1(vocab)

text = """"It's the last he painted, you know," Mrs. Gisburn said with pardonable pride."""
ids = tokenizer.encode(text)
print(ids)
decoded_text = tokenizer.decode(ids)
print(decoded_text)
# Below is an example of the how the text looks after encoding/converting to its corresponding numbers.

[1, 56, 2, 850, 988, 602, 533, 746, 5, 1126, 596, 5, 1, 67, 7, 38, 851, 1108, 754, 793, 7]
" It' s the last he painted, you know," Mrs. Gisburn said with pardonable pride.


In [12]:
# Let's apply it to some new text
text = "Hello, do you like tea?"
print(tokenizer.encode(text))
# If a word/token is not present in the dictionary, then it will fail

KeyError: 'Hello'

Basically, it is important to have a very large and diverse corpus, so that more words can be in the vocabulary.

# Lets modify the Tokenizer to include unknown words.
- Always have a way to incorporate new words into the vocabulary.

In [13]:
all_tokens = sorted(list(set(preprocessed)))
all_tokens.extend(["<|endoftext|>", "<|unk|>"])
vocab = {token: integer for integer, token in enumerate(all_tokens)}
print(len(vocab.items()))
# added 2 new tokens. One is for end_of_text and the other for unknown.

1132


In [14]:
for i, item in enumerate(list(vocab.items())[-5:]):
    print(item)

('younger', 1127)
('your', 1128)
('yourself', 1129)
('<|endoftext|>', 1130)
('<|unk|>', 1131)


In [15]:
class SimpleTokenizerV2:
    '''
    This is similar to class Simpletokenizerv1 except that it can also handle unknown tokens.
    '''
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}

    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [item.strip() for item in preprocessed if item.strip()]
        preprocessed = [item if item in self.str_to_int else "<|unk|>" for item in preprocessed]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    
    def decode(self, ids):
        text = ' '.join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

# Compare V1 and V2
- The class V1 is able to handle new words that are not in the dictionary.

In [16]:
text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."
text = " <|endoftext|> ".join((text1, text2))
print(text)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the palace.


In [17]:
tokenizer = SimpleTokenizerV2(vocab)
print(tokenizer.encode(text))

[1131, 5, 355, 1126, 628, 975, 10, 1130, 55, 988, 956, 984, 722, 988, 1131, 7]


In [18]:
print(tokenizer.decode(tokenizer.encode(text)))

<|unk|>, do you like tea? <|endoftext|> In the sunlit terraces of the <|unk|>.


- The reason for <|unk|> above is because the original piece of text does not contain the words `Hello` and `palace`.

# Byte Pair Encoding

In [19]:
from importlib.metadata import version
import tiktoken
print(f"tiktoken version: {version('tiktoken')}")

tiktoken version: 0.8.0


In [20]:
tokenizer = tiktoken.get_encoding("gpt2")

In [21]:
text = "Hello, do you like tea? <|endoftext|> In the sunlit terraces of the someunknownPlace."
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 286, 262, 617, 34680, 27271, 13]


In [22]:
# Below is the break down of word someunknownPlace. Note that the space before the word `some` is considered part of the work (` some`)
print(tokenizer.encode(" someunknownPlace"))
print(tokenizer.encode(" some"))
print(tokenizer.encode("unknown"))
print(tokenizer.encode("Place"))


[617, 34680, 27271]
[617]
[34680]
[27271]


In [23]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the someunknownPlace.


In [24]:
text = "Akwirw ier"
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[33901, 86, 343, 86, 220, 959]
Akwirw ier


# Create train and test

In [25]:
# open and encode the text with tiktoken
with open("data/chapter2/the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [27]:
enc_sample = enc_text[50:]

In [29]:
# Create sample of context 4. 
# 4 tokens as input and the 5th token is output.
context_size = 4
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287]
y:      [4920, 2241, 287, 257]


In [30]:
# here is an example of the sample generated for training
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, '------>', desired)

[290] ------> 4920
[290, 4920] ------> 2241
[290, 4920, 2241] ------> 287
[290, 4920, 2241, 287] ------> 257


In [50]:
# repeat but with actual words

In [31]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), '------>', tokenizer.decode([desired]))

 and ------>  established
 and established ------>  himself
 and established himself ------>  in
 and established himself in ------>  a


# Efficient data loading

In [61]:
import torch
from torch.utils.data import Dataset, DataLoader

In [62]:
class GPTDatasetV1(Dataset):
    def __init__(self, txt, tokenizer, max_length, stride):
        """
        Class for loading data
        txt: This is the text to be encoded
        tokenizer: This is the tokenizer to use
        max_length: This is the max length of the input sequences
        stride: This is tokens to skip when creating input sequences.
        """
        self.input_ids = []
        self.target_ids = []

        token_ids = tokenizer.encode(txt)

        for i in range(0, len(token_ids)-max_length, stride):
            input_chunk = token_ids[i:i+max_length]
            target_chunk = token_ids[i+1: i+1+max_length]
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))

    def __len__(self):
        return len(self.input_ids)
    
    def __getitem__(self, idx):
        return self.input_ids[idx], self.target_ids[idx]

In [63]:
def create_dataloader_v1(txt, batch_size=4, max_length=256, stride=128, shuffle=True, drop_last=True, num_workers=0):
    """
    Function to create an efficient dataloader. It serves as an iterator to generate data without having to load all the data in memory.
    Args:
    - txt: The text to use for creating the training examples
    - batch_size: create batches of what size
    - max_length: length of each training sample
    - stride: The stride to take when creating samples, so that samples are not continuous
    - shuffle: Shuffle the data or not every epoch
    - drop_last: Set to drop the last sample if the last sample would be incomplete
    - num_workers: number of workers to use
    """
    tokenizer = tiktoken.get_encoding('gpt2')
    dataset = GPTDatasetV1(txt, tokenizer, max_length, stride)
    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        drop_last=drop_last,
        num_workers=num_workers
    )
    return dataloader

In [69]:
dataloader = create_dataloader_v1(
    raw_text, batch_size=1, max_length=4, stride=1, shuffle=False
)

data_iter = iter(dataloader)
first_batch = next(data_iter)
print(first_batch)

[tensor([[  40,  367, 2885, 1464]]), tensor([[ 367, 2885, 1464, 1807]])]


- The `first_batch` contains tensors for the input_ids and target_ids.
- They are of size 4 since we set the max_length to that value.
- The meaning of stride = 1 can be seen below by fetching the next batch. Basically, the next batch is shifted by 1.
- Also batch_size is 1, so only 1 input is created.

In [70]:
second_batch = next(data_iter)
print(second_batch)

[tensor([[ 367, 2885, 1464, 1807]]), tensor([[2885, 1464, 1807, 3619]])]


# Another example

In [71]:
dataloader = create_dataloader_v1(raw_text, batch_size=8, max_length=4, stride=4)

data_iter = iter(dataloader)
inputs, targets = next(data_iter)

print("\nInputs:\n", inputs)
print("\nTargets:\n", targets)


Inputs:
 tensor([[16857,   262,  4469,   286],
        [  683,  1760,   416,   257],
        [  319,  3619,   438,   505],
        [  314,  4137,  3589,   284],
        [  502,    13,   198,   198],
        [  672,   270,  2838,     1],
        [  290,   319,   465, 11914],
        [  198,    38,   271, 10899]])

Targets:
 tensor([[  262,  4469,   286,   607],
        [ 1760,   416,   257, 38378],
        [ 3619,   438,   505,   286],
        [ 4137,  3589,   284,  1234],
        [   13,   198,   198,     1],
        [  270,  2838,     1,   319],
        [  319,   465, 11914,    11],
        [   38,   271, 10899,   550]])


- Note that the stride is 4. 
- This will utilize the entire dataset and will not skip a single word.
- It will also avoid any overlap between batches to prevent overfitting.

# TOKEN EMBEDDINGS
- The last step is to convert the tokens to embedding vectors.

In [87]:
# Example
input_ids = torch.tensor([2, 3, 5, 1])
vocab_size = 6
output_dim = 3

torch.manual_seed(123)

embedding_layer = torch.nn.Embedding(vocab_size, output_dim)
print(embedding_layer.weight)

Parameter containing:
tensor([[ 0.3374, -0.1778, -0.1690],
        [ 0.9178,  1.5810,  1.3010],
        [ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-1.1589,  0.3255, -0.6315],
        [-2.8400, -0.7849, -1.4096]], requires_grad=True)


In [89]:
print(embedding_layer(torch.tensor([3])))
# The value is the same as we have in the embedding matrix for 3.

tensor([[-0.4015,  0.9666, -1.1481]], grad_fn=<EmbeddingBackward0>)


# NOW on all the vocab_size

In [108]:
vocab_size = 50257 # vocab size for GPT2
output_dim = 256
token_embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

In [109]:
token_embeddings = token_embedding_layer(inputs)
print(inputs)
print(token_embeddings)

tensor([[16857,   262,  4469,   286],
        [  683,  1760,   416,   257],
        [  319,  3619,   438,   505],
        [  314,  4137,  3589,   284],
        [  502,    13,   198,   198],
        [  672,   270,  2838,     1],
        [  290,   319,   465, 11914],
        [  198,    38,   271, 10899]])
tensor([[[-1.7754e-02,  2.3365e-01, -1.6188e+00,  ..., -1.1135e+00,
          -2.0446e+00,  1.2343e+00],
         [-3.9616e-01,  5.5931e-01,  3.4120e+00,  ...,  6.1461e-01,
          -3.9812e-01,  7.9994e-01],
         [-4.0972e-01, -1.8490e+00, -7.9382e-01,  ...,  6.9640e-01,
           9.0896e-01, -1.1008e+00],
         [-5.5941e-01,  1.2612e+00, -9.6166e-01,  ..., -8.8642e-01,
          -4.2649e-02, -2.1107e+00]],

        [[ 6.8685e-01, -3.0238e+00, -3.1724e-01,  ...,  2.2242e-01,
           8.1771e-02, -2.7718e-01],
         [ 1.2500e-01, -1.0881e+00, -1.5740e+00,  ..., -5.5652e-01,
          -6.1905e-01,  8.5808e-01],
         [-1.1594e+00, -2.1946e-02,  9.3654e-01,  ...,  4.3338e

In [113]:
print(token_embeddings.shape)

torch.Size([8, 4, 256])


- In the example above, the batch_size was 8, and max_length of an input is 4. Each of the input token is encoded in 256 dimensions (embedding size)
- For the GPT model's absolute embedding approach, we just need to create another embedding layer that has the same dimension as the token_embedding_layer

In [124]:
max_length = 4
dataloader = create_dataloader_v1(
    raw_text,
    batch_size=8,
    max_length=max_length,
    stride=max_length,
    shuffle=False
)
data_iter = iter(dataloader)
inputs, targets = next(data_iter)
print("Inputs:\n ",inputs)
print("Input shape: \n", inputs.shape)

token_embeddings = token_embedding_layer(inputs)
print("Token Embeddings: \n", token_embeddings.shape)

# Here we are taking the absolute embedding approach as used in GPT.
# Requires to create another embedding layer with the same dimensions as the original one.
context_length = max_length
pos_embedding_layer = torch.nn.Embedding(context_length, output_dim)
pos_embeddings = pos_embedding_layer(torch.arange(context_length))
print("Position Embeddings: \n", pos_embeddings.shape)


Inputs:
  tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])
Input shape: 
 torch.Size([8, 4])
Token Embeddings: 
 torch.Size([8, 4, 256])
Position Embeddings: 
 torch.Size([4, 256])


In [126]:
input_embeddings = context_length + pos_embeddings